torch/_functorch/aot_autograd.py

```diff
@@ -1476,42 +1588,88 @@ def forward(ctx, *deduped_flat_tensor_args):
+            return tuple(raw_returns)
 
         @staticmethod
-        def backward(ctx, *all_flat_args):
+        def backward(ctx, *flat_args):
             # Calling convention: we expect a grad_out passed to the backward:
-            # - for every output of the fw that does *not* alias an input
-            # - for every updated_input generated by the fw that does *not* alias an input
-            # - for every size/stride metadata value for aliased outputs.
-            #   These are returned by the forward, but we just drop them in the backward.
-            #   We need to return them in the forward, but unfortunately there's no way to specify
-            #   in autograd.Function that certain non-tensor forward outputs shouldn't show up in the backward.
-            expected_grad_outs = CompiledFunction.num_non_aliased_outs + CompiledFunction.num_mutated_data_inputs
-            if CompiledFunction.num_aliasing_metadata_outs > 0:
-                flat_args = all_flat_args[:-CompiledFunction.num_aliasing_metadata_outs]
-                metadata_args = all_flat_args[-CompiledFunction.num_aliasing_metadata_outs:]
-                # metadata args are all ints/symints, which autograd will send Nones for as grad_outputs in the bw
-                assert all([x is None for x in metadata_args])
-                # delete
-                # for out_idx, (base_sizes, base_strides, base_storage_offset) in CompiledFunctions.fw_out_base_metadata.items():
+            # - for every output of the fw that does *not* alias an input or graph intermediate
+            # - for every updated_input generated by the fw that does *not* alias an input (aka only data-mutations)
+            # - for every graph intermediate that we need to use to generate an output later.
+            # The other outputs in the autograd.Function.forward that do *not* show up in the backward include:
+            # - outputs that alias inputs or graph intermediates
+            # - updated inputs due to metadata-only mutations.
+            # We need to return them in the forward, but ensure that they all do not get gradients in the backward,
+            # and we filter them out here before passing the remaining grad_outputs into the compiled backward.
+            expected_grad_outs = CompiledFunction.num_outputs + CompiledFunction.num_mutated_inputs
+            num_mutated_inps = CompiledFunction.num_mutated_inputs
 
+            assert len(flat_args) == expected_grad_outs
+            if CompiledFunction.num_mutated_metadata_only_inputs > 0 or CompiledFunction.num_outputs_aliased > 0:
+                inp_tangents, out_tangents = flat_args[0:num_mutated_inps], flat_args[num_mutated_inps:]
+                # input_info contains info on *every* input,
+                # But in the backward(), we are only given grad outputs for every mutated input.
+                # We then need to filter out the grad outputs that correspond to metadata-only mutations.
```

Comment:
Wait, no, silently filtering out grad outputs sounds wrong. If we don't support gradients on it, we need to error if downstream sends us a gradient, asserting that downstream sends us None. If you want it to be usable in requires grad compute downstream without inducing gradients, that output should have been requires_grad=False.

Reply:
We talked offline about this, but just to state the reasoning:

* It is problematic to return an output in an `autograd.Function.forward` that aliases an input, because autograd.Function needs to conservatively assume that is cannot be mutated (but user code might later mutate that output).
* Instead, we completely regenerate aliases, in the wrapper code outside of the `autograd.Function.apply()`.
* This same logic applies to input metadata mutations: When we functionalize, inplace-views on inputs will turn into  ordinary views, which we return as outputs in the `autograd.Function.forward()`. We then use the metadata from that output to apply the inplace-view directly to the input, with a `.as_strided_()` call